stdlib/public/core/SmallString.swift

```diff
@@ -213,14 +220,19 @@ extension _SmallString {
   }
 
   // Overwrite stored code units, including uninitialized. `f` should return the
-  // new count.
+  // new count. This will re-establish the invariant after `f` that all bits
+  // between the last code unit and the discriminator are unset.
   @inline(__always)
-  internal mutating func withMutableCapacity(
+  fileprivate mutating func withMutableCapacity(
     _ f: (UnsafeMutableRawBufferPointer) throws -> Int
   ) rethrows {
     let len = try withUnsafeMutableBytes(of: &self._storage) {
       (rawBufPtr: UnsafeMutableRawBufferPointer) -> Int in
-      return try f(rawBufPtr)
+      let len = try f(rawBufPtr)
+      UnsafeMutableRawBufferPointer(
+        rebasing: rawBufPtr[len...]
+      ).initializeMemory(as: UInt8.self, repeating: 0)
```

Comment:
@atrick is this a proper use of `initializeMemory`? It's basically a convenient way to mask off our unused bytes and make sure they're set to 0.

Reply:
@milseman Now that I'm looking at this again, I _think_ this needs to rebind the memory; otherwise, after initializing memory as `UInt8`, portions of the memory are bound simultaneously as `RawBitPattern.self` and `UInt8.self` 😕 

```swift
let count = self.count
let len = try withUnsafeMutableBytes(of: &self._storage) {
  (rawBufPtr: UnsafeMutableRawBufferPointer) -> Int in
  let len = try f(rawBufPtr)
  rawBufPtr.bindMemory(to: UInt8.self, capacity: count)
  defer {
    _ = rawBufPtr.bindMemory(to: RawBitPattern.self, capacity: 1)
  }
  UnsafeMutableRawBufferPointer(
    rebasing: rawBufPtr[len...]
  ).initializeMemory(as: UInt8.self, repeating: 0)
  return len
}
```

Reply:
I don't see any API for initializing raw bytes without using a typed pointer. You can do the silly rebinding thing above or just write your own raw initialize that doesn't care about the in-memory type
```
var nextPtr = self
    for _ in 0..<count {
      Builtin.initialize(repeatedValue, nextPtr._rawValue)
      nextPtr += MemoryLayout<T>.stride
    }
```